Logs API to be user-facing (#4225)
open-telemetry/opentelemetry-specification: specification/logs/api.md

    + [Enabled](#enabled)
- [Optional and required parameters](#optional-and-required-parameters)
- [Concurrency requirements](#concurrency-requirements)
- [Convenience](#convenience)
- [Logs Bridge API](#logs-bridge-api)
- [References](#references)

<!-- tocstop -->

</details>

## Overview

This document defines a log API.

The API serves following use cases.

It is provided for logging library authors to build
[log appenders/bridges](./supplementary-guidelines.md#how-to-create-a-log4j-log-appender),
which use this API to bridge between existing logging libraries and the
OpenTelemetry log data model. See also: [Logs Bridge API](#logs-bridge-api)

It is provided for (instrumentation) libraries to emit

> [OpenTelemetry Events](data-model.md#events).

Reviewer: I believe by making Otel Logs API user-facing, we should not limit instrumentation to events. It should be fine to emit nameless log records. 

So maybe 

```suggestion
logs.
```

?

Reviewer 2: > I believe by making Otel Logs API user-facing, we should not limit instrumentation to events. It should be fine to emit nameless log records.
> 
> So maybe
> 
> ?

ABSOLUTELY!

And whether that are called "nameless" or there is some other "name" (not `event.name`) that describes that `type` (options discussed include `log.name`, `log.record.name` maybe even a new `log.type` (But the enumeration of this is extremely problematic as what "level" do you define)) of Log entry is specifically being written is just part of the unanswered questions.

Author: It is about OTel instrumentation libraries (not any "instrumentation").

> When OpenTelemetry instrumentation emits logs, those logs SHOULD be formatted as Events. 

from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md#event-data-model

as well as https://github.com/open-telemetry/oteps/blob/58d6420237091bcfb739f329e09adaa7ec03f702/text/0265-event-vision.md#L29-L32

The next paragraph is about using the API to emit any logs.

Reviewer 2: And both of those links talked about them be "represented" by the same Data Model, or as a "type" of Log, it is NOT stating that this makes them the SAME thing (beyond the transport / representation), they still HAVE different Semantics.
Circling back to the OTep this is the HOW are they SENT not WHAT are they.

What is an Event https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md
What is a Log https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/logs.md

Nowhere is it saying that a Log is an Event

Reviewer 2: And as an example of additional cases where I'm saying that we are talking past each other on the possible solutions, 
Please go and see my comment from Sept 9th IN THE OTEP
https://github.com/open-telemetry/oteps/pull/265#issuecomment-2339337199, in the design section and the "Combining API's" (the exact thing you are trying to define here)

Nothing over there was "simply" ranting, it was about trying to inform and provide workable solutions that will work for ALL of US and the community.

Author: @lmolkova PTAL https://github.com/open-telemetry/opentelemetry-specification/pull/4225/commits/d3aefef7b4f62448e4d4302fdba9833691afa484

Reviewer 2: Oh and a specific call out on this

> It is about OTel instrumentation libraries (not any "instrumentation").

Are we or are we not defining an Open Common API for Everyone -- not JUST "OTel instrumentation libraries", at least that is what I'm trying to do -- https://opentelemetry.io/docs/what-is-opentelemetry/

> Focused on the generation, collection, management, and export of telemetry. A major goal of OpenTelemetry is that you can easily instrument your applications or systems, no matter their language, infrastructure, or runtime environment. The storage and visualization of telemetry is intentionally left to other tools.

With the mission / vision of  https://opentelemetry.io/community/mission/

> OpenTelemetry’s Mission: to enable effective observability by making high-quality, portable telemetry ubiquitous.

Reviewer: Let's get back to the original comment - we're making log API user-facing so that OTel instrumentations or user applications could emit logs or events using it directly. 

There could be many reasons to use OTel log api directly (bridge overhead, lack of structured logging, not being able to record structured body with 3rd party API,  ambiguity in how bridge maps properties from 3rd party logging facade to otel). There is nothing specific about events here, so I'm suggesting not to mention them in the motivation to behind user-facing logging API/

Author: > There is nothing specific about events here, so I'm suggesting not to mention them in the motivation to behind user-facing logging API/.

The reason I mentioned them is that currently the specification says that OpenTelemetry instrumentation libraries SHOULD emit well-defined events (and not just random log records). I think it would be bad if an instrumentation library emit log records that are not defined in https://github.com/open-telemetry/semantic-conventions. Or maybe we anticipate that the semantic conventions will define conventions for log records which will not be events?

I think also that this description is in the spirit of the OTEP: https://github.com/open-telemetry/oteps/blob/58d6420237091bcfb739f329e09adaa7ec03f702/text/0265-event-vision.md?plain=1#L29-L32

Reviewer: Imagine I'm instrumenting kafka:
- there could be some events I'd emit - e.g. cluster leader is unreachable - which are documented in the semconv
- there could also be things I just want to let my users know of via logs. E.g. certain configuration was picked, connection has dropped, etc. I could use 3rd party logging library, but why if I can use OTel logging API without adding a new dependency?

The library logs are emitted with is not relevant to users - they'd get the same logs if lib used 3rd party logger or OTel API. So I don't see why we should limit instrumentations from using OTel API to emit undocumented regular (but structured) logs.

Reviewer: > How would the instrumentation library report a problem if the OTel Logs Provider is not configured?

it won't - I mean someone could configure default provider (stdout), but in general case if you use 3rd party logging library, it's not a lib responsibility to notify you that your logging provider is not configured. 

We can borrow some ideas from logging libs. E.g. SLF4J notifies you if provider is not configured with something like this in stderr:

<img width="809" alt="image" src="https://github.com/user-attachments/assets/11118058-7daf-444a-98e1-b24c67607b6c">

--------------------

I agree that it's not a viable choice for instrumentation today to rely on OTel logs being configured as a main source of logging, but 3 years from now it could be a nice option to report all telemetry with otel instead of taking dependency on something else.  

And we're not restricting anything - we're adding more options.

In any case, I'd like to have this discussion tomorrow on the spec call, but if you feel strongly about making progress on this PR without addressing it, I'm fine doing it as a follow up.

Author: I created https://github.com/open-telemetry/opentelemetry-specification/issues/4234

Reviewer 3: > Imagine I'm instrumenting kafka:
> 
> - there could be some events I'd emit - e.g. cluster leader is unreachable - which are documented in the semconv
> 
> - there could also be things I just want to let my users know of via logs. E.g. certain configuration was picked, connection has dropped, etc. I could use 3rd party logging library, but why if I can use OTel logging API without adding a new dependency?
> 
> 
> 
> The library logs are emitted with is not relevant to users - they'd get the same logs if lib used 3rd party logger or OTel API. So I don't see why we should limit instrumentations from using OTel API to emit undocumented regular (but structured) logs.

I think this is a good example. You can emit an untyped event (which is a nameless logrecord), and it would bypass any event pipelines set up, but it would still give you a single api surface at the low level.

Reviewer: BTW @pellared my main focus with this is native instrumentations where now I need to pick one API for logging and use OTel for everything else. I.e. I don't mean instrumentation logs, I mean the underlying library logs.

    + [Enabled](#enabled)
- [Optional and required parameters](#optional-and-required-parameters)
- [Concurrency requirements](#concurrency-requirements)
- [References](#references)

<!-- tocstop -->

</details>

## Overview

> This document defines a log API.

Reviewer 2: As part of OTep #265 it was determined that there SHOULD be a public facing "Logging" API to enable the emitting of "Events" (as a type of OpenTelemetry Log) as well as other general "Logs" <- I'm implicitly inferring based on discussions that occurred as part of #265 (ie. while all `events` can be represented as Logs, not all Logs are `events`) 

One of the key words here is `type`, there are many `types` of Logs and each type fundamentally defines the "intent" (or the "why") the log is being emitted.

These can include "Event", "Audit", "Security", "Transaction", "Message", "Server", "File" and many, many more, and just as "Events are a `type` of Log", several of these can also be sub-classified as a type of `event`.

As part of any Logging API we SHOULD provide a front-end API and guidance for all of the CORE fundamental concepts to "try" to make it clear about what will and won't occur when you do or don't use the specific API. 
(Right NOW we don't have this clearly defined (because we are always trying to take small steps) which always lands us into the argument of "But it's just a Log", which is not a productive argument as it ignores the requirements by different teams / use cases that there are different needs and different `types` of Logs that should have some uniform structure (Semantic Conventions / processing) around them.

Author: I changed the PR so that  Logs API can be used to emit events but I have not removed the Events API.

> This document defines a log *backend* API.

Reviewer 4: I think part of the decision in the OTEP is to remove "backend" (by calling it a user-facing Logs API)

```suggestion
This document defines a log API.
```

> The API serves following use cases.

Reviewer 2: There are many additional features / requirements that exist and are yet to be defined as we move this API forward (some of which was starting to occur in the Event API) as it was moving (or trying to move) towards stability.
Some of these will include different processing requirements for the different `types` and as it moves towards a stabilized state it may cause implementation changes, so for now (as a small step) I'll try to define a starting point.

Author: I think this should rather require refining the SDK and not the API.

It is provided for instrumentation libraries to emit events (log records following

> OpenTelemetry Semantic Conventions).

Reviewer 2: It should include an `emitEvent` function which MUST include the `name` as a first-class concept as the specific value of the `name` defines the Semantic-Definition of any provided `body` for the specific `event`.

As the `name` is a required argument / property for this function it MUST also enforce that the `name` conforms to the OpenTelemetry general semantic conventions (ie. it is fully qualified -- just like an Attribute name (this doesn't mean a complete list of valid "registered" events), e.g. it just needs to have a leading domain and then name `domain.eventName` where the domain may have multiple separators.

And it should also
 - Be used only to emit a well-structured, well defined `event` as per the definitions from an OpenTelemetry and application perspective.
  - These `events` will include the fundamental log `types` like "Application", "User", "Client", "Device", "Security".
  - Which ["It signals that "this thing happened at this time" and provides additional specifics about the occurrence"](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/events.md#semantic-conventions-for-events)
  - Should be used by Applications (Adopters) and Instrumentations only.
  - Because the exact structure and semantic meaning of the `body` of the `event` is driven and defined by the `name`, this allows for SDK's (or applications) to extend and provide (if required)
    - explicit validation of the content
    - explicit sampling / rules governing that specific "instance"
    - explicit delivery mechanism (Send NOW, Don't Drop, Create Metric, as some possible examples)
    - The exact location of the "where" and "when" of any rules is up to the specific application / implementation (eg. it "MAY" occur on the client, in the SDK, at the aggregation point (Collector) or at rest and queried as required.

And (if) future versions of OpenTelemetry evolve to change the internal representation of the `event` as a `LogRecord` all of the "changes" can be limited to the SDK implementation/version of the API, so that all existing users of the API have zero change when it's just an internal "representation" within the `LogRecord` (not-withstanding that this would have downstream implementations)

> It is provided for logging library authors to build

Reviewer 2: - `emitBridgeLog` (just making the name different from above -- needs a better one), just here to have the distinction explicit.
  - Exists to "bridge" into the OpenTelemetry `Logs` pipeline only
  - Used by log Appenders ONLY to create OpenTelemetry `LogRecords` and have them processed
  - NO special processing (beyond existing Log Processors ??)*
    - (*)There may be some special use cases where some Log Processors may be required to be skipped (like a processor that links OTel Logs -> Native Logging)
    - Should we change this and have "explicit" Appender/Bridge Log Processors
  - IF an existing language / logging library supports the creation of "Events" or specific "Structured" Logs then this API MAY be used to "Simulate" (generate) a `LogRecord` that looks like those created by the above API's.
    - HOWEVER, this WILL NOT re-apply / perform any special processing that may be configured by the SDK instance for the other API methods
    - So it's entirely 100% up to the appender (caller of this API) to ensure that it is complying with any necessary semantic conventions to avoid any downstream complications around deterministically identifying / processing the resulting `LogRecord`.
    - It also MUST NOT just convert everything into an Event UNLESS
      - it can EXPLICITLY identify that the intent of the `LogRecord` was to identify (signal) that the specifically named `event` occurred vs general Detailed Logging
      - or it create a general semantic convention that identifies that this `LogRecord` that looks like an `event` was created by or follows the semantic conventions of X. ie. it has a specific `name` like `log.*`, `applog.*` or `event.ilogger` again so the intent (type) can be deterministically evaluated AND it would avoid any "clashes" with defined semantic conventions.
      - To phrase another way as long as the Appender "Knows" that the user emitting the `LogRecord` WANTS this specific  log to be called `XXXXX` then this is 100% acceptable as long as the control (setting) of this name is explicitly defined by the user. eg. Application has their own logger for their own events, but they want to explicitly `emit` an OpenTelemetry defined `event` with a Specific name via their own logger and not via the OpenTelemetry `emitEvent` with the understanding (implicit or otherwise) that any configured OpenTelemetry "Event" processing will NOT be performed via this path.

Reviewer 4: @MSNev can you highlight any specific blocking concerns you have about this PR beyond https://github.com/open-telemetry/opentelemetry-specification/pull/4225#issuecomment-2379899009, and open tracking issues for the rest? Once we get this initial restructuring PR landed then we'll be able to start sending PRs to try to address those. Thanks.

Author: It is up to the bridge implementation how the logging library would be mapped to OTel Logs Data Model. I think we should rather create some "guidance" / "recommendations" for log bridge authors. We cannot control how the bridges are going to be implemented.

For sure this should be out of scope of this PR.

[log appenders/bridges](./supplementary-guidelines.md#how-to-create-a-log4j-log-appender),
which use this API to bridge between existing logging libraries and the
OpenTelemetry log data model.

> It is provided for application developers to emit structured log records

Reviewer 2: This should call out the "difference" between the processing / handling / semantics of an `event` vs a structured log

- `emitLog`
  - Used to emit a structured Log (without (necessarily) any implicit meaning
  - These `logs` will include type "Server", "Transaction", "Message", "Audit", "Debug", "Diagnostic" etc.
  - To be used to provide "Detailed Information" which may be used to understand operational issues, often used for post mortem analysis rather than operational dashboards or monitoring.
    - Should be used by Applications, Instrumentations and SDK implementations.
  - If there is any explicit `event` processing, this API won't perform any explicit processing beyond general common Log Processors

Problem Statement and confusion
 - As "structured" logs generally also need some form of "identifier" about the "type", so these will also need some form of `name`, `schema` or `identifying` attributes as these (vs general unstructured logging) generally have some semantic conventions on their structure and beyond just have "OTLP" `AnyValues` processors will likely want a simple mechanism for identifying them.
 - The MUST however, NOT be identified in the same manner as `events`, as then there is no deterministic mechanism for identifying "How" the type of `LogRecord` was produced (as an Event or as a Detailed Information record).
 - As such should this include or be restricted to have a general (optional) `name`
 - If a Language / runtime has a standard logging library with a supported log appender then there will be confusion between using this API vs their specific (preferred) logging library.
   - The difference should be isolated around configured specific Log Processing (if any) -- which of course doesn't exist today.

Author: > This should call out the "difference" between the processing / handling / semantics of an event vs a structured log

I find it as out of scope. We need such mechanism for all log records. From last SIG meeting agenda:

> We need filtering and separate log processing pipelines. E.g. distinct processing for different exporters or different processing for events and other log records. Also needed for [Specify how Logs SDK implements Enabled #4207](https://github.com/open-telemetry/opentelemetry-specification/issues/4207)

Reviewer 2: Am I stating anywhere that we don't -- NO.
But I am stating that we ALSO need this for Events (The Event API has this for that exact purpose), we could have "just" ignored it and said hey go just use a genera Log processor on your Log definition.

Reviewer: based on the current version of spec, it's totally valid to emit events using log bridge. We don't have a requirement to support separate pipelines today and this PR does not change status quo 

https://github.com/open-telemetry/opentelemetry-specification/blob/fbaf8468c599d858df515c88bac58d02c9afdb29/specification/logs/event-api.md?plain=1#L79-L81

> It is provided for application developers to emit structured log recors

Reviewer 5: ```suggestion
It is provided for application developers to emit structured log records
```

> It is provided for application developers to emit structured log records
> (e.g. custom events).

Reviewer 4: ```suggestion
It is provided for application developers to emit events.
```

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/4225/files#r1777996355 + they can still emit any log record. The application developers can create their own semantic log records which they may want to be different than events.

Reviewer 4: ```suggestion
It is provided for application developers to emit structured log records
(including OpenTelemetry Events).
```

> Languges MAY provide a separate Logs Bridge API if they need different

Reviewer 6: ```suggestion
Languages MAY provide a separate Logs Bridge API if they need different
```

> ergonimics for consumers that are building log appenders/bridges.

Reviewer 5: ```suggestion
ergonomics for consumers that are building log appenders/bridges.
```

> The Logs API consist of these main components:

Reviewer 6: ```suggestion
The Logs API consists of these main components:
```

## Concurrency requirements

For languages which support concurrent execution the Logs APIs provide
specific guarantees and safeties.

**LoggerProvider** - all methods are safe to be called concurrently.

**Logger** - all methods are safe to be called concurrently.

> ## Convenience

Reviewer 2: This is PART of the issue with this process.

## Events are MORE than just a Convenience!

I would be fine with wording along the lines of

For languages that do not need or want to provide any OpenTelemetry Event support because they have their own native or common libraries, then they MAY choose to not implement the `emitEvent` functionality and support.

Author: > Events are MORE than just a Convenience

I find this as an opinionated statement based on the current definition of Opentelemetry Events

Reviewer 2: > I find this as an opinionated statement based on the current definition of Opentelemetry Events

## And THAT is the crux of the disagreement

You (and several others) DON'T believe and either DON'T understand or WANT to understand that "Events" are NOT JUST Logs, so your Opinionated view / statement is to remove something that you DON'T understand.

As stated in the Oel (this portion I agree with)

> OpenTelemetry Events are a **type** of OpenTelemetry Log that requires an event name and follows a specific structure implied by that event name.

They are a TYPE of Log, it does not state that ALL Logs are Events and it does not state the there is ONLY 1 TYPE of Log (Events). Calling Every LOG and `Event` destroys the Semantic meaning of what is an `Event`.

That does not mean that there isn't room to "define" the appropriate Semantics to keep Both, but that is the discussions and part of the agreements that we need to make BEFORE we can conclude this PR (as I've mentioned eariler)

Reviewer: Could we call this section "Optional methods" so it works for everyone?

Author: I would rather remove the whole section. 

I think it is needed not only for sake of events but e.g. Java may want to have convenient methods like `Info`, `Warn`, `Error`.

> Calling Every LOG and Event destroys the Semantic meaning of what is an Event.

Nobody does it. Currently, just call every log with `event.name` attribute as an Event. For the SDK/Data Model/Collector perspective it does not matter how it was created. This is currently the way an Event is defined and I am not making any change to it.

Author: Another problem, I see in defining `Emit an Event` method (in scope of this PR) is that there would be an ambiguity what should happen if a user adds an additional `event.name` attribute. Should the `name` parameter or `event.name` attribute have precedence?

PS. The same problem currently exists  in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md#emit-event

Author: I created https://github.com/open-telemetry/opentelemetry-specification/issues/4233

And also  committed https://github.com/open-telemetry/opentelemetry-specification/pull/4225/commits/cc0f26289786620e9708a15f7534347c777cb9b9 to avoid this issue in scope of this PR

> Languages MAY provide additional ergonomics and convinence APIs. For instance,

Reviewer 4: ```suggestion
Languages MAY provide additional ergonomics and convenience APIs. For instance,
```
